### PR TITLE
Multiple Improvements

### PR DESCRIPTION
**Pull request by liushuyu**
This pull request makes multiple improvements to the scripted GitHub Actions in this repository.

- Switch to recommended Node.js 18 TypeScript configuration provided by TypeScript upstream
- Updated the NPM packages to their latest versions to fix several security issues
- Replaced `@vercel/ncc` with `esbuild` for better bundling results (better tree-shaking and better build times)
- Use proper private member sigil in the code (see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_properties)
- Switch to Node.js 20 runtime to suppress deprecation warnings

### CODE REVIEW

**Comment by liushuyu** [comment]
Note that the generated files under `dist/` have not been modified. It's recommended that the person who merges this pull request generate the files on their own device to ensure safety.

File: rockcraft-pack/action.yml [hunk 1]
```diff
@@ -24,7 +25,7 @@ outputs:
   rock:
     description: 'The file name of the resulting ROCK.'
 runs:
-  using: 'node16'
+  using: 'node20'
```
Comment (cjdcordeiro) [inline rockcraft-pack/action.yml:28]: the GH workflow used `node18`. Is it a problem to have `node20` here? Why is it different?

Comment (liushuyu) [reply rockcraft-pack/action.yml:28]: Done. Now addressed in test.yml

**Review by cjdcordeiro — APPROVED** [review]
LGTM. We'll need someone from the @canonical/starcraft-maintainers to review it as well

**Review by sergiusens — APPROVED** [review]
thanks for the improvements

**Pull request merged**